### DISCUSSION TITLE
How to translate texts in ternary operator?

### DISCUSSION THREAD

**Discussion by crazyyi** [lingui/js-lingui#1507] [Q&A]
I have tried following this PR:
https://github.com/lingui/js-lingui/pull/1436

`<Trans>{isA? <Trans id="a">A</Trans> : isB ? <Trans id="b">B</Trans> : <Trans id="c">C</Trans>}</Trans>`
But it only generates translations with the same text as the id of the `Trans` tag
I got the same result with or without the outermost `Trans` tag

**Comment by timofei-iatsenko** [comment]
But what you expecting in outermost Trans? If you don't have any text to translate there it would be empty. 

Look at the test case to understand how macro transform your code: 

```ts
  {
    name: "JSX Macro inside JSX conditional expressions",
    stripId: true,
    input: `
       import { Trans } from '@lingui/macro';
       <Trans>Hello, {props.world ? <Trans>world</Trans> : <Trans>guys</Trans>}</Trans>
      `,
    expected: `
        import { Trans } from '@lingui/react';
       
        <Trans
          id={'<stripped>'}
          message={"Hello, {0}"}
          values={{
            0: props.world ? <Trans id={"<stripped>"} message={'world'} /> : <Trans id={'<stripped>'} message={'guys'} />
          }}
        />
      `,
  },
```

In catalog in this case you will get 3 messages: 
- `Hello, {0}` (outer Trans)
- `world`
- `guys`

If you don't want to split your message into chunks you can use Select format like so

```tsx
import {Select} from "@lingui/macro"
<Trans>Hello, <Select value="props.world ? 'world' : 'guys'" _world="World" _guys="Guys" /></Trans>
```

In catalog you receive 

`Hello {select, =world {World} =guys {Guys}}`

For ternary you can use `true` / `false` strings as keywords for Select expression

**Reply by crazyyi** [reply]
Thanks. I mean I get "a" : "a", "b" : "b", "c": "c" in my json file.
Maybe I should try the Select syntax. in the latter example, does `'world' : 'guys'` use just as symbols? And 
`_world="World" _guys="Guys" ` is used for assigning translation texts to the symbols?

**Reply by timofei-iatsenko** [reply]
> Thanks. I mean I get "a" : "a", "b" : "b", "c": "c" in my json file.

Comparing to your input, output looks fine. 

Here is reference for Select macro
https://lingui.dev/ref/macro#select-1

**Reply by crazyyi** [reply]
But I was expecting "a": "A", "b": "B", "c": "C". How can I do that?

**Reply by timofei-iatsenko** [reply]
don't use explicit id's, it's highly discouraging

**Reply by crazyyi** [reply]
> don't use explicit id's, it's highly discouraging

Isn't an explicit id easier for translators to understand the purpose and location of the texts?

**Reply by timofei-iatsenko** [reply]
no

**Reply by timofei-iatsenko** [reply]
Use po files. It has origins automatically extracted which show where this translation is used.  

The approach with explicit ids is not scalable. You have "naming things" problem multiplied by thy amount of your messages. You will have collision problems earlier or later.  You will have a lot of duplicates in catalogs which you have to translate manually.

Just write this `<Trans>Hello {name}!</Trans>` or 

```
t`Hello {name}`
```

it's much more natural, cleaner and scales well. I know that this bullshit for explicit ids is written in the docs. The docs already updated for v4
